Fix: #43 Update World Data Dictionary

dr5hn opened the pull request:
# This PR includes
- Updated world data dictionary to the latest and less erroneous.

@harpreetkhalsagtbit 
Note: With this PR the Upcoming release will become Backward Incompatible.

Suggestion, 
- We can also add a Country flag emoji if you want.

harpreetkhalsagtbit commented:
this contains Ids. I thought we are getting rid of Ids.

dr5hn commented:
@harpreetkhalsagtbit

Are you sure?

I don't think getting rid of ID's is good for cities. It is playing an important role in unique identifiers for cities.

harpreetkhalsagtbit commented:
Ok, we can keep Ids for the cities. But it would be better if instead of proprietor Ids, we use standard Country, State codes and remove Ids at all if possible.

dr5hn commented:
@harpreetkhalsagtbit 

Nice idea !! We can do this.

Well, then you need to make changes in your code as well to handle this.

harpreetkhalsagtbit commented:
> @harpreetkhalsagtbit
> 
> Nice idea !! We can do this.
> 
> Well, then you need to make changes in your code as well to handle this.

Changing code is not a big issue. But updating the data is. I am not getting much time reviewing your database. The new json structure should be

```js
// Old
{
        "id": "1",
        "sortname": "AF",
        "name": "Afghanistan",
        "phonecode": "93"
}

// new

{
        "isoCode": "AF",
        "name": "Afghanistan",
        "phonecode": "93"
},

```

dr5hn commented:
@harpreetkhalsagtbit

Sure I will work on it this weekend.

harpreetkhalsagtbit commented:
Hi,
Please share your email and contact no.
I will try to spare some time tomorrow for this. Ideally, I would like to spend a whole day on this. But I would be busy with some personal stuff. 
Let's plan things out and I would provide full support whatever would be possible for me.
Thanks

dr5hn commented:
@harpreetkhalsagtbit 
FYI, Tomorrow I will be working on this in my evenings.

Email: gadadarshan[at]gmail[dot]com
Phone: +91 9[zero]292[zero]8698

dr5hn commented:
@harpreetkhalsagtbit Done!

- Updated Data Structure as per your requirements.
- Added a flag
- Update code to some extent support new data structure

Note: I have not tested the code, but I have done suitable updates to the code.

dr5hn commented:
@harpreetkhalsagtbit 

I have added a GitHub action to run tests on pull requests 
- It will start working when you push the `ci.yml` file to master.
- This action will trigger when someone raises a PR it will run a build and tests on that PR to verify for correctness. 
- For your reference, You can check here I have raised a PR on the forked repo and it runs a test on that PR. ([PR](https://github.com/dr5hn/country-state-city/pull/1)) ([Test Check Suite](https://github.com/dr5hn/country-state-city/pull/1/checks?check_run_id=1581768147))

dr5hn commented on README.md:
```diff
@@ -164,9 +168,14 @@ Special Thanks
 Change Logs
 ---------------
+**v2.0.0 (Backward Incompatible)** :: [Data Source (dr5hn)](https://github.com/dr5hn/countries-states-cities-database)
 
-**v1.0.0**
+1. ``
+2. ``
+3. ``
+4. ``
```
@harpreetkhalsagtbit Let me know what to update here?

dr5hn commented:
@harpreetkhalsagtbit
Did you get a chance to look into it?

dr5hn commented:
@harpreetkhalsagtbit Reminder.

harpreetkhalsagtbit commented:
Great work @dr5hn!

More things that we can add in the future would be:
1. Add currency and telephone code for each country. We can also contribute back to the original database.
2. National Language of the country.
3. GPS coordinates for Country. If possible(States, Cities).
4. Time zone - GMT

harpreetkhalsagtbit commented:
I think we should rename - `countryCode` to actual `iso3`.

dr5hn commented:
@harpreetkhalsagtbit My DB is connected with the iso2 code for states and cities. 

So what's the next task?

harpreetkhalsagtbit commented:
> Great work @dr5hn!
> 
> More things that we can add in the future would be:
> 
> 1. Add currency and telephone code for each country. We can also contribute back to the original database.
> 2. National Language of the country.
> 3. GPS coordinates for Country. If possible(States, Cities).
> 4. Time zone - GMT

Let's finish this first. I have added a few comments for the current PR. Please finish them first. Then I have mentioned the future plan in one of the comments above, which is not a big thing, it is already there in your DB.
@dr5hn

harpreetkhalsagtbit reviewed (commented):
> I think we should rename - `countryCode` to actual `iso3`. But that would mean a lot of changes in the code wherever we are using field - countrycode

Let's ignore this for now. We will discuss this later.

dr5hn commented on index.ts:
```diff
@@ -5,55 +5,53 @@ import { ICountry, ICity, IState} from './src/interface';
+    },
+    getStateByCode: function (isoCode: string): IState {
+        return _findEntryByCode(stateList, isoCode);
+    }
```
@harpreetkhalsagtbit `getCityByCode(countryCode, stateCode)` is already there. Since we're passing Country Code and State Code, it matches with `getCitiesOfState(countryCode, stateCode)`.

So do we need to replicate?

harpreetkhalsagtbit replied:
`getCitiesOfState(countryCode, stateCode)` will return array of cities. I want city details with city code as input.

dr5hn replied:
@harpreetkhalsagtbit 

We cannot get a single city with `countryCode` and `stateCode` as an input. We need one more identifier to get a single city. For ex: it can be a `cityName`

harpreetkhalsagtbit replied:
leave it for now..I will check this later....

dr5hn commented:
> 1. Add currency and telephone code for each country. We can also contribute back to the original database.
> 2. National Language of the country.
> 3. GPS coordinates for Country. If possible(States, Cities).
> 4. Time zone - GMT

@harpreetkhalsagtbit 
- I have already added a phone code to the current code. 
- I have currency, co-ordinates & timezone in my DB, so I will add them.
- I don't have the National language of the country. Is that fine?

dr5hn commented:
@harpreetkhalsagtbit Done. Please verify.

harpreetkhalsagtbit commented:
> > 1. Add currency and telephone code for each country. We can also contribute back to the original database.
> > 2. National Language of the country.
> > 3. GPS coordinates for Country. If possible(States, Cities).
> > 4. Time zone - GMT
> 
> @harpreetkhalsagtbit
> 
> * I have already added a phone code to the current code.
> * I have currency, co-ordinates & timezone in my DB, so I will add them.
> * I don't have the National language of the country. Is that fine?

No issues for national language lets add what we have now.

harpreetkhalsagtbit commented:
I want to understand how to use `"gmtOffset": 16200,` value to calculate time? what exacatly is this value `16200`? Its in seconds/milliseconds?

dr5hn commented:
@harpreetkhalsagtbit Already added the remainings. Please verify.

harpreetkhalsagtbit commented:
If nothing left, then I can merge this. I will check and test this at the weekend.
@dr5hn Thanks a lot!

harpreetkhalsagtbit commented:
Thank you so much. I have merged this into my `develop` branch. Now preparing for master release and cross check existing open issues and hopefully it will help me close many of them.
@dr5hn
